uploaders/common.go

```diff
@@ -110,32 +110,60 @@ func createArtifact(buildURL, token, artifactPth, artifactType string) (string,
 func uploadArtifact(uploadURL, artifactPth, contentType string) error {
 	log.Printf("uploading artifact")
 
-	data, err := os.Open(artifactPth)
-	if err != nil {
-		return fmt.Errorf("failed to open artifact, error: %s", err)
+	netClient := &http.Client{
+		Timeout: 10 * time.Minute,
 	}
-	defer func() {
-		if err := data.Close(); err != nil {
-			log.Errorf("Failed to close file, error: %s", err)
+
+	return retry.Times(3).Wait(5).Try(func(attempt uint) error {
+		file, err := os.Open(artifactPth)
+		if err != nil {
+			return fmt.Errorf("failed to open artifact, error: %s", err)
 		}
-	}()
+		defer func() {
+			if err := file.Close(); err != nil {
+				log.Warnf("failed to close file, error: %s", err)
+			}
+		}()
 
-	args := []string{"curl", "--fail", "--tlsv1", "--globoff"}
-	if contentType != "" {
-		args = append(args, "-H", fmt.Sprintf("Content-Type: %s", contentType))
-	}
-	args = append(args, "-T", artifactPth, "-X", "PUT", uploadURL)
+		request, err := http.NewRequest(http.MethodPut, uploadURL, ioutil.NopCloser(file))
+		if err != nil {
+			return fmt.Errorf("failed to create request, error: %s", err)
+		}
 
-	return retry.Times(3).Wait(5 * time.Second).Try(func(attempt uint) error {
-		if attempt > 0 {
-			log.Warnf("%d attempt failed", attempt)
+		request.Header.Add("Content-Type", contentType)
+
+		// Set Content Length manually (https://stackoverflow.com/a/39764726), as it is part of signature in signed URL
+		fileInfo, err := file.Stat()
+		if err != nil {
+			return fmt.Errorf("failed to get file info for %s, error: %s", artifactPth, err)
 		}
-		cmd, err := command.NewFromSlice(args)
+		request.ContentLength = fileInfo.Size()
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+		defer cancel()
+		request = request.WithContext(ctx)
+
+		resp, err := netClient.Do(request)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to upload artifact, error: %s", err)
 		}
-		cmd.SetStdout(os.Stdout).SetStderr(os.Stderr)
-		return cmd.Run()
+
+		defer func() {
+			if err := resp.Body.Close(); err != nil {
+				log.Errorf("Failed to close response body, error: %s", err)
+			}
+		}()
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read response body, error: %s", err)
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			return fmt.Errorf("invalid status code: %d, headers: %s, body: %s", resp.StatusCode, resp.Header, body)
```

Comment:
Note: status code is not invalid, if it is not ok, it indicates that the request failed for some reason. Invalid status code would be 9999 or something that is not commonly used or agreed beforehand.